lib/kbuildbarn/urls.go

```diff
@@ -0,0 +1,130 @@
+package kbuildbarn
+
+import (
+	"fmt"
+	"github.com/enfabrica/enkit/lib/multierror"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// ByteStreamUrl retieves the CAS id and bytes of an action based on the input url.
+// For example, bytestream://build.local.enfabrica.net:8000/blobs/a9a664559b4d29ecb70613fad33acfb287f2fa378178e131feaaebb5dafa231a/465
+// should return (a9a664559b4d29ecb70613fad33acfb287f2fa378178e131feaaebb5dafa231a, 465, nil)
+// which is a BuildBarnParams.Hash, BuildBarnParams.Size, error.Error
+func ByteStreamUrl(byteStream string) (string, string, error) {
+	u, err := url.Parse(byteStream)
+	if err != nil {
+		return "", "", err
+	}
+	splitUrl := strings.Split(u.Path, "/")
+	if len(splitUrl) != 4 {
+		return "", "", fmt.Errorf("ByteStreamUrl() bytestream url is not well formed %s", byteStream)
+	}
+	return splitUrl[2], splitUrl[3], nil
```

Comment:
Shouldn't this function also return the base URL that's needed by NewBuildBarnParams()? How else can the application get it from the original bytestream: URL?

Reply:
The Buildbarn base url is not necessarily what's in the bytestream URL.
For example, if you communicate over ssh to remote exec, the url in the bytestream url is 127.0.0.1. If it's via unix sockets the url is a file <.<. Neither are where the CAS is stored.